flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -75,10 +137,281 @@ public DefaultScheduler(
 			slotRequestTimeout,
 			shuffleMaster,
 			partitionTracker);
+
+		this.restartBackoffTimeStrategy = checkNotNull(restartBackoffTimeStrategy);
+		this.slotRequestTimeout = checkNotNull(slotRequestTimeout);
+		this.slotProvider = checkNotNull(slotProvider);
+		this.delayExecutor = checkNotNull(delayExecutor);
+		this.userCodeLoader = checkNotNull(userCodeLoader);
+		this.schedulingStrategyFactory = checkNotNull(schedulingStrategyFactory);
+		this.failoverStrategyFactory = checkNotNull(failoverStrategyFactory);
+		this.executionVertexOperations = checkNotNull(executionVertexOperations);
+		this.executionVertexVersioner = checkNotNull(executionVertexVersioner);
+		this.conditionalFutureHandlerFactory = new ConditionalFutureHandlerFactory(executionVertexVersioner);
 	}
 
+	// ------------------------------------------------------------------------
+	// SchedulerNG
+	// ------------------------------------------------------------------------
+
 	@Override
-	public void startScheduling() {
-		throw new UnsupportedOperationException();
+	public void startSchedulingInternal() {
+		initializeScheduling();
+		schedulingStrategy.startScheduling();
+	}
+
+	private void initializeScheduling() {
+		executionFailureHandler = new ExecutionFailureHandler(failoverStrategyFactory.create(getFailoverTopology()), restartBackoffTimeStrategy);
+		schedulingStrategy = schedulingStrategyFactory.createInstance(this, getSchedulingTopology(), getJobGraph());
+		executionSlotAllocator = new DefaultExecutionSlotAllocator(slotProvider, getInputsLocationsRetriever(), slotRequestTimeout);
+		setTaskFailureListener(new UpdateTaskExecutionStateInDefaultSchedulerListener(this, getJobGraph().getJobID()));
+		prepareExecutionGraphForScheduling();
+	}
+
+	@Override
+	public boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
+		final Optional<ExecutionVertexID> executionVertexIdOptional = getExecutionVertexId(taskExecutionState.getID());
+		if (executionVertexIdOptional.isPresent()) {
+			final ExecutionVertexID executionVertexId = executionVertexIdOptional.get();
+			updateState(taskExecutionState);
+			schedulingStrategy.onExecutionStateChange(executionVertexId, taskExecutionState.getExecutionState());
```

Comment:
Seems the schedulingStrategy is not aware of internal state changes in this way, like DEPLOYING or SCHEDULED, and can only trigger scheduling upon the external state changes, like RUNNING, FINISHED, etc. Some state changes like CANCELED or FAILED might be notified when it's from task manager and not notified from scheduler internal errors.

Reply:
> Seems the schedulingStrategy is not aware of internal state changes in this way, like DEPLOYING or SCHEDULED, and can only trigger scheduling upon the external state changes, like RUNNING, FINISHED, etc. 

That's a correct observation. At the moment states other than `FINISHED` are not needed but the inconsistency should be fixed eventually. The `DefaultScheduler` could invoke `SchedulingStrategy#onExecutionStateChange(...)` at the right places.

> Some state changes like CANCELED or FAILED might be notified when it's from task manager and not notified from scheduler internal errors.

`UpdateTaskExecutionStateInDefaultSchedulerListener` is supposed to propagate internal errors to `DefaultScheduler`.

Reply:
By design execution state changes should be notified via `SchedulingStrategy #onExecutionStateChange`.
Similarly, all partition consumable state changes should be notified via `SchedulingStrategy #onPartitionConsumable`, as discussed the in [another comment](https://github.com/apache/flink/pull/9663/files#r326540913)
This is how we define this public interface and I think we need to guarantee its correctness eventually.

And you are right that it is not a problem at the moment, or before we mark it as a public interface for users or other developers to customize the scheduling strategy. 
So I think we can skip it in this PR and do this in a follow up task.

Reply:
[FLINK-14233](https://issues.apache.org/jira/browse/FLINK-14233) is opened to track this.